tests/http_spec.rs

```diff
@@ -0,0 +1,215 @@
+    fn read(spec: &str) -> Option<Self> {
+      spec
+        .split('.')
+        .last()
+        .and_then(|ext| match ext.to_lowercase().as_str() {
+          "json" => Self::read_json(spec),
+          "yaml" => Self::read_yaml(spec),
+          _ => None,
+        })
+    }
+    fn read_json(spec: &str) -> Option<Self> {
+      let contents = fs::read_to_string(spec).ok()?;
+      let spec = serde_json::from_str(&contents);
+
+      spec.ok()
+    }
+
+    fn read_yaml(spec: &str) -> Option<Self> {
+      let contents = fs::read_to_string(spec).ok()?;
+      let spec = serde_yaml::from_str(&contents);
+      spec.ok()
+    }
+    async fn setup(&self) -> Arc<ServerContext> {
+      let config = Config::from_file_paths([self.config.clone()].iter())
+        .await
+        .ok()
+        .unwrap();
+      let blueprint = Blueprint::try_from(&config).unwrap();
+      let client = Arc::new(MockHttpClient { upstream_mocks: Arc::new(self.upstream_mocks.to_vec()) });
+      let server_context = ServerContext::new(blueprint, client);
+      Arc::new(server_context)
+    }
+  }
+
+  #[derive(Clone)]
+  struct MockHttpClient {
+    upstream_mocks: Arc<Vec<(UpstreamRequest, UpstreamResponse)>>,
+  }
+  #[async_trait::async_trait]
+  impl HttpClient for MockHttpClient {
+    async fn execute(&self, req: reqwest::Request) -> anyhow::Result<Response> {
+      let upstream_mocks = self.upstream_mocks.clone();
+      let upstream_mock = upstream_mocks.iter().find(|(upstream_request, _res)| {
+        let method_match = req.method().as_str()
+          == serde_json::to_string(&upstream_request.0.method.clone().unwrap_or_default())
+            .unwrap()
+            .as_str()
+            .trim_matches('"');
+        let url_match = req.url().as_str() == req.uri().to_string().as_str();
+        method_match && url_match
+      });
+      let upstream_response = upstream_mock.unwrap().clone().1.clone();
+      let mut response =
+        Response { status: reqwest::StatusCode::from_u16(upstream_response.0.status).unwrap(), ..Default::default() };
+      let headers = upstream_response.0.headers.unwrap_or_default();
+      for (k, v) in headers.iter() {
+        response.headers.insert(
+          HeaderName::from_str(k.as_str()).unwrap(),
+          HeaderValue::from_str(v.as_str()).unwrap(),
+        );
+      }
+      match upstream_response.0.body {
+        None => {
+          response.body = ConstValue::Null;
+        }
+        Some(a) => {
+          response.body = ConstValue::try_from(serde_json::from_value::<Value>(a).unwrap())?;
+        }
+      }
+      Ok(response)
+    }
+  }
+
+  #[tokio::test]
+  async fn test_body_yaml() {
+    let spec = HttpSpec::read("tests/data/sample.yaml").unwrap();
+    assert_downstream(spec).await;
+  }
+
+  async fn assert_downstream(spec: HttpSpec) {
+    for downstream_assertion in spec.clone().downstream_assertions.iter() {
+      if let Some(annotation) = downstream_assertion.annotation.clone() {
```

Comment:
adding additional default variant to annotation could simplify using it here

Reply:
simplified

Reply:
default is not needed we not annotation is needed case only can run